File: src/938. Range Sum of BST/main.go
// Solved by @kitanoyoru

package main

type TreeNode struct {
  Val int
  Left *TreeNode
  Right *TreeNode
}

func dfs(node *TreeNode, low, high int, ans *int) {
  if node == nil {
    return
  }
  if node.Val >= low && node.Val <= high {
    *ans += node.Val
  }
  
  if node.Val > low {
    dfs(node.Left, low, high, ans)
  }
  if node.Val < high {
    dfs(node.Right, low, high, ans)
  }
}

func rangeSumBST(root *TreeNode, low int, high int) int {
  var ans int
  dfs(root, low, high, &ans);
  return ans
}

File: src/938. Range Sum of BST/main.py
# Solved by @kitanoyoru

from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def dfs(self, node: Optional[TreeNode]) -> None:
        if node is None:
            return

        if node.val >= self.low and node.val <= self.high:
            self.ans += node.val

        if node.val > self.low:
            self.dfs(node.left)
        if node.val < self.high:
            self.dfs(node.right)

    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        self.ans: int = 0
        self.low = low
        self.high = high

        self.dfs(root)

        return self.ans
